### PR TITLE
Generate footpaths

### PR DESCRIPTION
**Pull request by Tristramg**

### CODE REVIEW

File: src/structures.rs [hunk 1]
```diff
@@ -33,47 +38,60 @@ impl Timetable {
             .parse::<NaiveDate>()
             .expect("Could not parse start date");
 
+        let stops: Vec<_> = gtfs.stops
+            .iter()
+            .map(|stop| {
```
Comment (antoine-de) [inline src/structures.rs]: even if it will be a bit more verbose, I think it's better to implement the `From` trait

File: src/gtfs_structures.rs [hunk 2]
```diff
@@ -49,8 +49,8 @@ pub struct CalendarDate {
 pub struct Stop {
     #[serde(rename = "stop_id")] pub id: String,
     pub stop_name: String,
-    location_type: Option<u8>,
-    parent_station: Option<String>,
+    pub location_type: Option<u8>,
```
Comment (antoine-de) [inline src/gtfs_structures.rs]: what is the location_type ?

File: src/structures.rs [hunk 3]
```diff
@@ -109,4 +127,38 @@ impl Timetable {
...
         result
     }
+
+    fn footpaths(stops: &Vec<Stop>, stop_indices: &HashMap<String, u32>) -> Vec<Vec<Footpath>> {
+        let mut result: Vec<Vec<_>> = stops.iter().map(|_| Vec::new()).collect();
+        let mut stop_areas = HashMap::new();
+
+        for stop in stops {
+            if let Some(ref parent) = stop.parent_station {
+                if stop.location_type == 1 || stop.location_type == 0 {
```
Comment (antoine-de) [inline src/structures.rs]: smell like `location_type` should be an enum 😜

File: src/structures.rs [hunk 4]
```diff
@@ -109,4 +127,38 @@ impl Timetable {
...
+                let index_a = *stop_indices
+                    .get(child_a)
+                    .expect(&format!("Missing child station {}", child_a));
+                let index_b = *stop_indices
+                    .get(child_b)
+                    .expect(&format!("Missing child station {}", child_b));
+
+                if index_b != index_a {
+                    result[index_a as usize].push(Footpath {
+                        duration: 5,
```
Comment (antoine-de) [inline src/structures.rs]: why this `5` ?

File: src/structures.rs [hunk 5]
```diff
@@ -109,4 +127,38 @@ impl Timetable {
...
+        for (parent, children) in stop_areas {
+            for (child_a, child_b) in children.iter().cartesian_product(&children) {
+                let index_a = *stop_indices
+                    .get(child_a)
+                    .expect(&format!("Missing child station {}", child_a));
+                let index_b = *stop_indices
+                    .get(child_b)
+                    .expect(&format!("Missing child station {}", child_b));
+
+                if index_b != index_a {
```
Comment (antoine-de) [inline src/structures.rs]: really minor, but the check can be done before the indices get no ? (I think it add some meaning about what you want to test).
or add a `.filter(|a, b| a == b)` ?

File: src/gtfs_structures.rs (hunk 2 above)
Comment (TeXitoi) [reply src/gtfs_structures.rs]: https://developers.google.com/transit/gtfs/reference/#stopstxt

Comment (TeXitoi) [reply src/gtfs_structures.rs]: It can be a `u8` as the default value is 0

File: src/structures.rs [hunk 6]
```diff
@@ -18,13 +20,16 @@ struct Connection {
     arr_stop: u32,
 }
 
-struct Footpath {}
+struct Footpath {
+    to: u32,
+    duration: u16,
```
Comment (TeXitoi) [inline src/structures.rs:37]: u16 for duration will be small, no?

File: src/gtfs_structures.rs (hunk 2 above)
Comment (antoine-de) [reply src/gtfs_structures.rs]: my point was that it was easier to understand if it was an enum.

File: src/structures.rs (hunk 4 above)
Comment (Tristramg) [reply src/structures.rs]: minutes, I’ll fine tune that later (reading also `transfers.txt`)

File: src/structures.rs (hunk 6 above)
Comment (Tristramg) [reply src/structures.rs:37]: no, everything is in minutes, not seconds. Maybe I’ll regret that later, but for now I’ll just ignore

**Comment by Tristramg** [comment]
All comments have been considered. Thank you for the review

File: src/structures.rs [hunk 7]
```diff
@@ -33,22 +45,12 @@ pub struct Timetable {
...
-            .map(|stop| {
-                Stop {
-                    id: stop.id.to_owned(),
-                    name: stop.stop_name.to_owned(),
-                    parent_station: stop.parent_station.to_owned(),
-                    location_type: stop.location_type,
-                }
-            })
-            .collect();
+        let stops: Vec<_> = gtfs.stops.iter().map(|s| Stop::from(s)).collect();
```
Comment (TeXitoi) [inline src/structures.rs]: `.map(Stop::from).collect()`

Comment (Tristramg) [reply src/structures.rs]: indeed… thanks!

File: src/gtfs_structures.rs [hunk 8]
```diff
@@ -49,8 +56,8 @@ pub struct CalendarDate {
 pub struct Stop {
     #[serde(rename = "stop_id")] pub id: String,
     pub stop_name: String,
-    location_type: Option<u8>,
-    parent_station: Option<String>,
+    #[serde(deserialize_with = "deserialize_location_type")] pub location_type: LocationType,
```
Comment (antoine-de) [inline src/gtfs_structures.rs:59]: rustfmt let you write this on the same line ?

**Review by antoine-de — APPROVED** [review]

File: src/gtfs_structures.rs (hunk 8 above)
Comment (Tristramg) [reply src/gtfs_structures.rs:59]: It actually puts it in one line

**Pull request merged**